File: StripeGate/src/main/java/com/androidafe/stripegate/StripeGate.java
/*
 *
 *   Created Md Rahadul Islam on 12/16/23, 3:16 PM
 *   Copyright Ⓒ 2023. All rights reserved Ⓒ 2023 http://freefuninfo.com/
 *   Last modified: 12/16/23, 3:04 PM
 *
 *   Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file
 *   except in compliance with the License. You may obtain a copy of the License at
 *   http://www.apache.org/licenses/LICENS... Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
 *    either express or implied. See the License for the specific language governing permissions and
 *    limitations under the License.
 * /
 *
 */

package com.androidafe.stripegate;

import android.content.Context;
import android.os.Handler;
import android.util.Log;
import android.widget.Toast;

import androidx.activity.ComponentActivity;
import androidx.annotation.NonNull;

import com.androidafe.stripegate.models.CustomerResponse;
import com.androidafe.stripegate.models.EphemeralResponse;
import com.androidafe.stripegate.models.PaymentResponse;
import com.androidafe.stripegate.modules.PARAMS;
import com.androidafe.stripegate.modules.ProgressDialog;
import com.androidafe.stripegate.modules.RetrofitClient;
import com.stripe.android.PaymentConfiguration;
import com.stripe.android.paymentsheet.PaymentSheet;
import com.stripe.android.paymentsheet.PaymentSheetResult;
import com.stripe.android.paymentsheet.PaymentSheetResultCallback;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class StripeGate {

    private PaymentSheet paymentSheet;
    private String customerID, ephemeralKey, clientSecret, amount, currency;
    Context context;
    ProgressDialog dialog;

    private static String SECRET_KEY;
    private static String PUBLISHABLE_KEY;
    private boolean paymentSuccessful;
    private PaymentResultListener paymentResultListener;
    public static void setSecretKey(String secretKey) {
        SECRET_KEY = secretKey;
    }

    public static void setPublishableKey(String publishableKey) {
        PUBLISHABLE_KEY = publishableKey;
    }


    public StripeGate(Context context) {
        this.context = context;
        initPaymentSheet();

        // Check if PUBLISHABLE_KEY and SECRET_KEY are not null
        if (PUBLISHABLE_KEY == null || SECRET_KEY == null) {
            Log.d("error", "Publishable Key or Secret Key is empty: ");
        }

         dialog = new ProgressDialog(context);
    }


    private void initPaymentSheet() {
        PaymentSheetResultCallback paymentSheetCallback = new PaymentSheetResultCallback() {
            @Override
            public void onPaymentSheetResult(@NonNull PaymentSheetResult paymentSheetResult) {
                StripeGate.this.onPaymentResult(paymentSheetResult);
            }
        };

        paymentSheet = new PaymentSheet((ComponentActivity) context, paymentSheetCallback);
    }



    public void Integrate(String amount, String currency) {
        if (PUBLISHABLE_KEY == null || SECRET_KEY == null) {
            Log.e("error", "Publishable Key or Secret Key is empty ");
            return; // Return early if keys are null
        }
        this.amount = amount;
        this.currency = currency;
        PaymentConfiguration.init(context, PUBLISHABLE_KEY);
        createCustomerID();
        getPaymentInfo(this.amount, this.currency); // Use class fields here
    }

    // Overloaded method to accept double for the amount
    public void Integrate(double amount, String currency) {
        Integrate(String.valueOf((long) (amount * 100)), currency);
    }


    public void Apply(final PaymentResultListener paymentResultListener) {

        if (PUBLISHABLE_KEY == null || SECRET_KEY == null) {
            Log.d("error", "Publishable Key or Secret Key is empty ");
            paymentResultListener.onPaymentResult(false);
            return; // Return early if keys are null
        }

        dialog.show();
        createCustomerID();
        getPaymentInfo(amount,currency);
        new Handler().postDelayed(new Runnable() {
            @Override
            public void run() {
                dialog.dismiss();
                PaymentFlow();
            }
        },2000);

    }



    private void createCustomerID(){
        Call<CustomerResponse> call = RetrofitClient.getInstance().getAPI().getCustomerID("Bearer "+ SECRET_KEY);
        call.enqueue(new Callback<CustomerResponse>() {
            @Override
            public void onResponse(Call<CustomerResponse> call, Response<CustomerResponse> response) {
                if (response.isSuccessful()){
                    CustomerResponse cm = response.body();
                    customerID = cm.getId();
                    getEphemeralKey(customerID);
                }else {
                    Log.d("res", "onResponse: failed");
                }
            }

            @Override
            public void onFailure(Call<CustomerResponse> call, Throwable t) {
                Log.d("res", "onError: failed"+t.getLocalizedMessage());
            }
        });
    }

    private void getEphemeralKey(String id){
        Call<EphemeralResponse> keys = RetrofitClient.getInstance().getAPI().getEphemeralKeys("Bearer "+SECRET_KEY
                ,"2023-10-16",customerID);

        keys.enqueue(new Callback<EphemeralResponse>() {
            @Override
            public void onResponse(Call<EphemeralResponse> call, Response<EphemeralResponse> response) {
                if (response.isSuccessful()){
                    EphemeralResponse ephemeralResponse = response.body();
                    ephemeralKey = ephemeralResponse.getEphemeralKey();

                }else {
                    Log.d("res", "onResponse: failed");
                }
            }

            @Override
            public void onFailure(Call<EphemeralResponse> call, Throwable t) {
                Log.d("res", "ephemeral: failed"+t.getLocalizedMessage());
            }
        });

    }

    public void getPaymentInfo(String amount, String currency){
        Call<PaymentResponse> paymentInfo = RetrofitClient.getInstance().getAPI().getPaymentInfo("Bearer "+SECRET_KEY,
                customerID, amount,currency,true);
        paymentInfo.enqueue(new Callback<PaymentResponse>() {
            @Override
            public void onResponse(Call<PaymentResponse> call, Response<PaymentResponse> response) {
                if (response.isSuccessful()){
                    PaymentResponse res = response.body();
                    clientSecret = res.getClientSecret();
                }else {
                    Log.d("res", "onError:"+response.errorBody());
                }
            }

            @Override
            public void onFailure(Call<PaymentResponse> call, Throwable t) {
                Log.d("res", "onError: failed"+t.getLocalizedMessage());
            }
        });
    }

    public void PaymentFlow() {
        if (ephemeralKey != null && clientSecret != null && customerID != null) {
            paymentSheet.presentWithPaymentIntent(
                    clientSecret, new PaymentSheet.Configuration("My Company",
                            new PaymentSheet.CustomerConfiguration(
                                    customerID,
                                    ephemeralKey
                            ))
            );
        } else {
            Log.d("res", "EphemeralKey, ClientSecret, or CustomerID is null. Cannot initiate payment flow.");
        }
    }

    public void onPaymentResult(PaymentSheetResult paymentSheetResult) {
        if (paymentSheetResult instanceof PaymentSheetResult.Completed){
            customerID = null;
            ephemeralKey = null;
            paymentSuccessful = true;
            Toast.makeText(context, "Payment Successfull", Toast.LENGTH_SHORT).show();
        }else {
            paymentSuccessful = false;
            Toast.makeText(context, "Payment Failed", Toast.LENGTH_SHORT).show();
        }
        if (paymentResultListener != null) {
            paymentResultListener.onPaymentResult(paymentSuccessful);
        }
    }

    public interface PaymentResultListener {
        void onPaymentResult(boolean isSuccessful);
    }

}

File: StripeGate/src/main/java/com/rahad/stripegate/StripeGate.kt
package com.rahad.stripegate

import android.content.Context
import android.util.Log
import androidx.activity.ComponentActivity
import androidx.lifecycle.MutableLiveData
import com.stripe.android.PaymentConfiguration
import com.stripe.android.paymentsheet.PaymentSheet
import com.stripe.android.paymentsheet.PaymentSheetResult
import com.stripe.android.paymentsheet.PaymentSheetResultCallback
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

class StripeGate(private val context: Context) {

    private val TAG = "StripeGate"

    private lateinit var paymentSheet: PaymentSheet
    private lateinit var customerId: String
    private var ephemeralKey: String? = null
    private var clientSecret: String? = null
    private var dialog: ProgressDialog

    private val _amount = MutableLiveData<String>()
    private val _currency = MutableLiveData<String>()
    private val SECRET_KEY = MutableLiveData<String>()
    private val PUBLISHABLE_KEY = MutableLiveData<String>()
    private var paymentSuccessful: Boolean = false

    private lateinit var paymentResultListener: PaymentResultListener

    private val systemAPI = ApiService.getApiInterface()

    init {
        dialog = ProgressDialog(context)
        initPaymentSheet()
    }

    fun setSecretKey(key: String) {
        SECRET_KEY.postValue(key)
    }

    fun setPublishableKey(key: String) {
        PUBLISHABLE_KEY.postValue(key)
        PaymentConfiguration.init(context, key)
    }

    private fun initPaymentSheet() {
        val paymentSheetCallback = PaymentSheetResultCallback { paymentSheetResult ->
            onPaymentResult(paymentSheetResult)
        }
        paymentSheet = PaymentSheet(context as ComponentActivity, paymentSheetCallback)
    }

    fun integrate(amount: String, currency: String) {
        _amount.postValue(amount)
        _currency.postValue(currency)
    }

    fun applyPayment(paymentResultListener: PaymentResultListener) {
        this.paymentResultListener = paymentResultListener
        if (PUBLISHABLE_KEY.value.isNullOrEmpty() || SECRET_KEY.value.isNullOrEmpty()) {
            Log.e(TAG, "Please setup publishable key and secret key")
            return
        }
        createCustomerId()
    }

    private fun createCustomerId() {
        CoroutineScope(Dispatchers.IO).launch {
            try {
                dialog.showOnMainThread()
                val response = systemAPI.getCustomerID("Bearer ${SECRET_KEY.value}")
                if (response.isSuccessful) {
                    customerId = response.body()!!.id
                    Log.d(TAG, "Customer ID created: $customerId")
                    getEphemeralKey(customerId)
                } else {
                    logErrorAndDismiss("Failed to create customer ID", response.errorBody()?.string())
                }
            } catch (e: Exception) {
                logExceptionAndDismiss("createCustomerId", e)
            }
        }
    }

    private fun getEphemeralKey(id: String) {
        CoroutineScope(Dispatchers.IO).launch {
            try {
                val response = systemAPI.getEphemeralKeys("Bearer ${SECRET_KEY.value}", "2024-06-20", id)
                if (response.isSuccessful) {
                    ephemeralKey = response.body()!!.secret
                    Log.d(TAG, "Ephemeral key received: $ephemeralKey")
                    getPaymentInfo(_amount.value!!, _currency.value!!, id)
                } else {
                    logErrorAndDismiss("Failed to get ephemeral key", response.errorBody()?.string())
                }
            } catch (e: Exception) {
                logExceptionAndDismiss("getEphemeralKey", e)
            }
        }
    }

    private fun getPaymentInfo(amount: String, currency: String, customerId: String) {
        CoroutineScope(Dispatchers.IO).launch {
            try {
                val response = systemAPI.getPaymentInfo(
                    token = "Bearer ${SECRET_KEY.value}",
                    customerID = customerId,
                    amount = amount,
                    currency = currency,
                    value = true
                )
                if (response.isSuccessful) {
                    clientSecret = response.body()!!.client_secret
                    Log.d(TAG, "Client secret received: ${response.body().toString()}")
                    paymentFlow()
                } else {
                    logErrorAndDismiss("Failed to get payment info", response.errorBody()?.string())
                }
            } catch (e: Exception) {
                logExceptionAndDismiss("getPaymentInfo", e)
            }
        }
    }

    private fun paymentFlow() {
        CoroutineScope(Dispatchers.Main).launch {
            try {
                if (clientSecret != null && ephemeralKey != null && ::customerId.isInitialized) {
                    paymentSheet.presentWithPaymentIntent(
                        clientSecret!!,
                        PaymentSheet.Configuration(
                            "Asian Group Distributor",
                            PaymentSheet.CustomerConfiguration(customerId, ephemeralKey!!)
                        )
                    )
                } else {
                    Log.e(TAG, "paymentFlow: Missing required data")
                    dialog.dismiss()
                }
            } catch (e: Exception) {
                logExceptionAndDismiss("paymentFlow", e)
            }
        }
    }

    private fun onPaymentResult(paymentSheetResult: PaymentSheetResult) {
        paymentSuccessful = paymentSheetResult is PaymentSheetResult.Completed
        paymentResultListener.onPaymentResult(paymentSuccessful)
        dialog.dismiss()
        if (paymentSuccessful) resetKeys()
    }

    private fun resetKeys() {
        customerId = ""
        ephemeralKey = ""
    }

    private fun logErrorAndDismiss(message: String, error: String?) {
        Log.e(TAG, "$message: $error")
        dialog.dismissOnMainThread()
    }

    private fun logExceptionAndDismiss(functionName: String, e: Exception) {
        Log.e(TAG, "Exception in $functionName: ${e.message}")
        dialog.dismissOnMainThread()
    }

    private fun ProgressDialog.showOnMainThread() {
        CoroutineScope(Dispatchers.Main).launch { show() }
    }

    private fun ProgressDialog.dismissOnMainThread() {
        CoroutineScope(Dispatchers.Main).launch { dismiss() }
    }

    interface PaymentResultListener {
        fun onPaymentResult(isSuccessful: Boolean)
    }
}
